src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Activity.cs

```diff
@@ -125,6 +126,95 @@ public async Task<IEnumerable<InstanceMetadata>> UpdateInstanceWatermarkV2Async(
+    public async Task<UpdateInstanceResponse> UpdateInstanceBlobsV3Async(
+        [ActivityTrigger] UpdateInstanceBlobArgumentsV2 arguments,
+        ILogger logger)
+    {
+        EnsureArg.IsNotNull(arguments, nameof(arguments));
+        EnsureArg.IsNotNull(arguments.ChangeDataset, nameof(arguments.ChangeDataset));
+        EnsureArg.IsNotNull(arguments.InstanceMetadataList, nameof(arguments.InstanceMetadataList));
+        EnsureArg.IsNotNull(arguments.Partition, nameof(arguments.Partition));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        DicomDataset datasetToUpdate = GetDeserialzedDataset(arguments.ChangeDataset);
+
+        int processed = 0;
+
+        logger.LogInformation("Beginning to update all instance blobs, Total count {TotalCount}", arguments.InstanceMetadataList.Count);
+
+        ConcurrentBag<InstanceMetadata> updatedInstances = new ConcurrentBag<InstanceMetadata>();
+        var errors = new ConcurrentBag<string>();
+
+        while (processed < arguments.InstanceMetadataList.Count)
+        {
+            int batchSize = Math.Min(_options.BatchSize, arguments.InstanceMetadataList.Count - processed);
+            var batch = arguments.InstanceMetadataList.Skip(processed).Take(batchSize).ToList();
+
+            logger.LogInformation("Beginning to update instance blobs for range [{Start}, {End}]. Total batch size {BatchSize}.",
+                    batch[0],
+                    batch[^1],
+                    batchSize);
+
+            await Parallel.ForEachAsync(
+                batch,
+                new ParallelOptions
+                {
+                    CancellationToken = default,
+                    MaxDegreeOfParallelism = _options.MaxParallelThreads,
+                },
+                async (instance, token) =>
+                {
+                    try
+                    {
+                        FileProperties fileProperties = await _updateInstanceService.UpdateInstanceBlobAsync(instance, datasetToUpdate, arguments.Partition, token);
+                        updatedInstances.Add(
+                            new InstanceMetadata(
+                                instance.VersionedInstanceIdentifier,
+                                new InstanceProperties
+                                {
+                                    FileProperties = fileProperties,
+                                    NewVersion = instance.InstanceProperties.NewVersion,
+                                    OriginalVersion = instance.InstanceProperties.OriginalVersion
+                                }));
+                    }
+                    catch (DataStoreRequestFailedException ex)
+                    {
+                        logger.LogInformation("Failed to update instance with watermark {Watermark}", instance.VersionedInstanceIdentifier.Version);
+                        errors.Add($"{ex.Message}. Instance UIDs - {instance.VersionedInstanceIdentifier}");
```

Comment:
Are we sure we want to output the `ToString` for `VersionedInstanceIdentifier` here?

Reply:
Updated to include Version. For esma's question, user is querying for an operation id. May be in shoebox it is helpful and not here may be